File: spring-decks/src/main/java/com/gpt/decks/engine/president/Rng.java
package com.gpt.decks.engine.president;

import java.util.List;

/**
 * Seedable mulberry32 PRNG — deterministic and serializable via its 32-bit
 * {@link #state()}, so a game can be persisted/replayed. A faithful port of the
 * TS engine's RNG; authority now lives here, so determinism is JVM-internal.
 */
public final class Rng {

  private int s;

  public Rng(int seed) {
    this.s = seed;
  }

  public static Rng fromState(int state) {
    return new Rng(state);
  }

  public int state() {
    return s;
  }

  /** Next float in [0, 1). */
  public double next() {
    s = s + 0x6d2b79f5; // int overflow wraps == TS's >>> 0 on the state bits
    int t = s;
    t = (t ^ (t >>> 15)) * (t | 1);
    t ^= t + (t ^ (t >>> 7)) * (t | 61);
    return ((t ^ (t >>> 14)) & 0xFFFFFFFFL) / 4294967296.0;
  }

  /** Integer in [0, maxExclusive). */
  public int nextInt(int maxExclusive) {
    return (int) (next() * maxExclusive);
  }

  /** Fisher–Yates shuffle in place. */
  public <T> void shuffle(List<T> list) {
    for (int i = list.size() - 1; i > 0; i--) {
      int j = nextInt(i + 1);
      T tmp = list.get(i);
      list.set(i, list.get(j));
      list.set(j, tmp);
    }
  }
}

File: angular-client/akira-ng/projects/president/src/app/game/rng.ts
/**
 * Tiny seedable PRNG (mulberry32). Deterministic and serializable via its 32-bit
 * {@link Rng.state}, so a server can persist/replay a game and clients can verify
 * shuffles. Not cryptographically secure — it only needs to be reproducible.
 */
export class Rng {
  private s: number;

  constructor(seed: number) {
    this.s = seed >>> 0;
  }

  /** Resume an RNG from a previously captured {@link state}. */
  static fromState(state: number): Rng {
    return new Rng(state);
  }

  /** Current internal state — store this in serializable game state. */
  get state(): number {
    return this.s >>> 0;
  }

  /** Next float in [0, 1). */
  next(): number {
    this.s = (this.s + 0x6d2b79f5) >>> 0;
    let t = this.s;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  }

  /** Integer in [0, maxExclusive). */
  int(maxExclusive: number): number {
    return Math.floor(this.next() * maxExclusive);
  }

  /** Fisher–Yates shuffle in place; returns the same array for convenience. */
  shuffle<T>(arr: T[]): T[] {
    for (let i = arr.length - 1; i > 0; i--) {
      const j = this.int(i + 1);
      [arr[i], arr[j]] = [arr[j], arr[i]];
    }
    return arr;
  }
}
